c++/src/kj/async.h

```diff
@@ -458,6 +458,21 @@ PromiseForResult<Func, void> evalLast(Func&& func) KJ_WARN_UNUSED_RESULT;
 // callback enqueues new events, then latter callbacks will not execute until those events are
 // drained.
 
+template <typename E, typename T, typename... Args>
+kj::Promise<T> runInEnvironment(E&& environment, FunctionParam<kj::Promise<T>(Args...)> func);
+// Synchronously calls `func`, arranging so that any calls to `getEnvironment()` from within sync
+// _or_ async user code will return a reference to an object constructed from `environment`.
+//
+// Note that Environments never implicitly cross threads.
+
+template <typename T>
+T& getEnvironment();
+template <typename T>
+kj::Maybe<T&> tryGetEnvironment();
+// If the currently-executing code was called by `runInEnvironment()`, or if the currently-executing
+// code was scheduled asynchronously by code called by `runInEnvironment()`, return the current
+// environment downcast as a value of type T.
```

Comment:
It feels concerning that if you try to get the environment as the wrong type here, you'll get UB (or an exception in debug mode). This sort of means that there has to be agreement among all code in a process that there is one true environment type. If that's the case then maybe we should be defining the type in KJ.

At the same time, I noticed that AsyncLocalContext in node and similar API proposals actually support multiple overlapping async contexts. Maybe we want to do the same here. I would key it by type. So if you have one set of code using environments of type T and another set using type U, calling each other, they won't interfere.

I think we can do this without RTTI. If you have a function like:

```
template <typename T>
void* getEnvironmentKey() {
  static const bool b;
  return &b;
}
```

This should return a unique value for each type.

Reply:
I like that idea, too, will give it a shot.